posts/gups/gups.cu

```diff
@@ -0,0 +1,598 @@
+#define NSHMEM 29056
+#else
+// not supported, fall back to dynamic shmem allocation
+#undef STATIC_SHMEM
+#endif
+
+#endif
+
+union benchtype
+{
+  uint64_t data;
+};
+static __constant__ uint64_t c_m2[64];
+
+#ifndef CUDA_RT_CALL
+#define CUDA_RT_CALL(call)                                                         \
+  {                                                                                \
+    cudaError_t cudaStatus = call;                                                 \
+    if (cudaSuccess != cudaStatus)                                                 \
+      fprintf(                                                                     \
+          stderr,                                                                  \
+          "ERROR: CUDA RT call \"%s\" in line %d of file %s failed with %s "       \
+          "(%d).\n",                                                               \
+          #call,                                                                   \
+          __LINE__,                                                                \
+          __FILE__,                                                                \
+          cudaGetErrorString(cudaStatus),                                          \
+          cudaStatus);                                                             \
+  }
+#endif
+
+static __global__ void d_init(size_t n, benchtype* t)
+{
+  for (ptrdiff_t i = blockIdx.x * blockDim.x + threadIdx.x; i < n;
+       i += gridDim.x * blockDim.x) {
+    t[i].data = i;
+  }
+}
+
+static __device__ uint64_t d_starts(size_t n)
+{
+  if (n == 0) {
+    return 1;
+  }
+
+  int i = 63 - __clzll(n);
+
+  uint64_t ran = 2;
+  while (i > 0) {
+    uint64_t temp = 0;
+    for (int j = 0; j < 64; j++) {
+      if ((ran >> j) & 1) {
+        temp ^= c_m2[j];
+      }
+    }
+    ran = temp;
+    i -= 1;
+    if ((n >> i) & 1) {
+      ran = (ran << 1) ^ ((int64_t)ran < 0 ? POLY : 0);
+    }
+  }
+
+  return ran;
+}
+
+enum test_t
+{
+  UPDATE = 0,
+  READ,
+  WRITE,
+  READ_WRITE,
+  UPDATE_NO_LOOP,
+};
+
+const char* test_name[] = {
+    "Gupdates/s ATOM.CAS.64(loop)",
+    "Greads/s LD.64",
+    "Gwrites/s STG.64",
+    "Greads+writes/s LDG.64+STG.64",
+    "Gupdates/s ATOM.CAS.64(no_loop)",
+};
+
+template <test_t T_TYPE>
+__global__ void
+d_bench(size_t n, size_t working_set, benchtype* t, int accesses_per_elem)
+{
+  size_t num_threads = gridDim.x * blockDim.x;
+  size_t thread_num = blockIdx.x * blockDim.x + threadIdx.x;
+  size_t start = thread_num * accesses_per_elem * n / num_threads;
+  size_t end = (thread_num + 1) * accesses_per_elem * n / num_threads;
+  benchtype ran;
+  ran.data = d_starts(start);
+
+#pragma unroll
+  for (ptrdiff_t i = start; i < end; ++i) {
+    ran.data = (ran.data << 1) ^ ((int64_t)ran.data < 0 ? POLY : 0);
+    unsigned long long int *address, old, assumed;
+    address = (unsigned long long int*)&t[ran.data & (working_set - 1)].data;
+    switch (T_TYPE) {
```

Comment:
Is this switch statement resolved at compile time? `T_TYPE` is a compile-time constant. However, there is no `switch constexpr` to force the compiler to resolve it. Alternatively, one can achieve this logic through template specialization, e.g. `template<> dbench<READ>(...)` or using a cascade of `if constexpr`, which is a C++17 feature.

Reply:
Yes, I believe the switch statement is resolved at compile time since all the `T_TYPE`s are known at compile time. I don't think template specification or cascade `if constexpr` is necessary...